scripts/install-build-tools.sh

```diff
@@ -14,30 +15,216 @@ if (( $EUID != 0 )); then
-  $SUDO apt install -y build-essential wget cmake libgtest-dev libbenchmark-dev lcov git software-properties-common rsync unzip
+  PYTHON_PATH=''
+  # check if supported version of python3 is already installed, and save the version
+  for PY_VERS in "${PYTHON_VERSIONS[@]}"; do
+    PYTHON_PATH="${BREW_ROOT}/bin/python${PY_VERS}"
+    # save the installed version if found
+    if [ -e "${PYTHON_PATH}" ]; then
+      PY_INSTALLED=${PY_VERS}
+      break
+    fi
+  done
 
-  wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | $SUDO apt-key add -
-  $SUDO add-apt-repository "deb http://apt.llvm.org/focal/ llvm-toolchain-focal-14 main"
+  # install valid python version if not installed yet
+  if [ -z "$PY_INSTALLED" ]; then
+    for PY_VERS in "${PYTHON_VERSIONS[@]}"; do
+      FULL_PY="python${PY_VERS}"
+      PYTHON_PATH="${BREW_ROOT}/bin/${FULL_PY}"
+      # try to install python version from homebrew and verify installation
+      if brew install "${FULL_PY}"; then
+        echo "${FULL_PY} installed successfully"
+        PY_INSTALLED=${PY_VERS}
+        break
+      else
+        # don't exit as we can try on another version if not the last one
+        echo "${FULL_PY} installation failed"
+      fi
+    done
+  fi
+
+  # leave if no valid python version is installed
+  if [ -z "$PY_INSTALLED" ]; then
+    echo "Python3 install with brew failed, attempted on these versions: ${PYTHON_VERSIONS[@]}"
+    exit 1
+  fi
+
+  # create virtual environment and install python packages for the valid python version
+  create_venv_install_python "${PYTHON_PATH}" ${PY_INSTALLED}
+
+  echo "To activate the virtual env to run python, run 'source .py_venv/bin/activate'"
+
+# Linux install with apt
+elif [[ "$OSTYPE" == "linux-gnu"* ]]; then
+  # avoids getting stuck on interactive prompts which is essential for CI/CD
+  export DEBIAN_FRONTEND=noninteractive
+  $SUDO apt update -y
+  $SUDO apt upgrade -y
+  $SUDO apt install -y build-essential wget cmake libgtest-dev libbenchmark-dev \
+    lcov git software-properties-common rsync unzip bc
+
+  # Add LLVM GPG key (apt-key is deprecated in Ubuntu 21.04+ so using gpg)
+  wget -qO - https://apt.llvm.org/llvm-snapshot.gpg.key | \
+    gpg --dearmor -o /usr/share/keyrings/llvm-archive-keyring.gpg
+  echo "deb [signed-by=/usr/share/keyrings/llvm-archive-keyring.gpg] http://apt.llvm.org/focal/ llvm-toolchain-focal-14 main" | \
+    $SUDO tee /etc/apt/sources.list.d/llvm.list
+
+  $SUDO apt update -y
   $SUDO apt install -y clang-format-14 clang-tidy-14
-  $SUDO ln -s -f $(which clang-format-14) /usr/local/bin/clang-format
-  $SUDO ln -s -f $(which clang-tidy-14) /usr/local/bin/clang-tidy
+  $SUDO ln -sf $(which clang-format-14) /usr/local/bin/clang-format
+  $SUDO ln -sf $(which clang-tidy-14) /usr/local/bin/clang-tidy
+
+  # check if supported version of python3 is already installed, and save the version
+  PYTHON_PATH=''
+  for PY_VERS in "${PYTHON_VERSIONS[@]}"; do
+    PYTHON_PATH="/usr/bin/python${PY_VERS}"
```

Comment:
Why not something like taking the output of `python3 --version` and trimming it? Python may not live at this path so we should avoid double installs

Reply:
Python3 could be linked to a version below 3.10 and we don't want to be limited to the default version. The goal was to avoid installing additional Python versions if we have valid ones already installed

Reply:
Seems like a reasonable alternative would be to try locating each version we're interested in with `which "python${PY_VERS}"` so that, regardless of where they're installed, if they're discoverable in-`$PATH`, we'll find them.

Reply:
The majority of my python versions are not discoverable in $PATH, but doing testing the specific version is a good strategy. I'll add that
```console
$ which python3.7
python3.7 not found
$ echo $?
1
$ which python3.12
/opt/homebrew/opt/python@3.12/bin/python3.12
$ echo $?
0
```

Reply:
If they're not discoverable in a well-known directory, or `$PATH`, we probably shouldn't care about them. Making a good-faith effort to be respectful of a user's system and space is always good; but, it's impossible to cover all edge-cases.